src/Features/CSharp/Portable/ForeachToFor/CSharpForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,160 @@
+
+            // create for statement from foreach statement
+            var forStatement = SyntaxFactory.ForStatement(
+                SyntaxFactory.VariableDeclaration(SyntaxFactory.IdentifierName("var"),
+                    SyntaxFactory.SingletonSeparatedList(
+                        SyntaxFactory.VariableDeclarator(
+                        SyntaxFactory.Identifier(indexString).WithAdditionalAnnotations(RenameAnnotation.Create()),
+                        default,
+                        SyntaxFactory.EqualsValueClause(
+                            SyntaxFactory.LiteralExpression(SyntaxKind.NumericLiteralExpression, SyntaxFactory.Literal(0)))))),
+                SyntaxFactory.SeparatedList<ExpressionSyntax>(),
+                SyntaxFactory.BinaryExpression(SyntaxKind.LessThanExpression, SyntaxFactory.IdentifierName(indexString), SyntaxFactory.ParseExpression($"{collectionVariableName}.{foreachInfo.CountName}")),
+                SyntaxFactory.SingletonSeparatedList<ExpressionSyntax>(SyntaxFactory.PostfixUnaryExpression(SyntaxKind.PostIncrementExpression, SyntaxFactory.IdentifierName(indexString))),
+                bodyStatement);
+
+            // let leading and trailing trivia set
+            if (!foreachInfo.RequireCollectionStatement)
+            {
+                forStatement = forStatement.ReplaceToken(forStatement.ForKeyword, forStatement.ForKeyword.WithLeadingTrivia(foreachStatement.ForEachKeyword.LeadingTrivia));
+            }
+
+            forStatement = forStatement.ReplaceToken(forStatement.CloseParenToken, forStatement.CloseParenToken.WithTrailingTrivia(foreachStatement.CloseParenToken.TrailingTrivia));
+
+            editor.ReplaceNode(foreachStatement, forStatement);
+        }
+
+        private ExpressionSyntax GetForeachCollection(SyntaxNode node)
+        {
+            var foreachStatement = (ForEachStatementSyntax)node;
+
+            // use original text, not value text
+            return foreachStatement.Expression;
+        }
+
+        private StatementSyntax GetForLoopBody(ForeachInfo foreachInfo, ParseOptions parseOption, string collectionVariableName, string indexString)
+        {
+            var foreachStatement = (ForEachStatementSyntax)foreachInfo.ForeachStatement;
+            if (foreachStatement.Statement is EmptyStatementSyntax)
+            {
+                return foreachStatement.Statement;
+            }
+
+            var bodyBlock = foreachStatement.Statement is BlockSyntax block ? block : SyntaxFactory.Block(foreachStatement.Statement);
+            if (bodyBlock.Statements.Count > 0)
+            {
+                // use original text
+                var foreachVariableString = foreachStatement.Identifier.ToString();
+
+                // create varialbe statement
+                var variableStatement = AddElasticAnnotation(SyntaxFactory.ParseStatement($"var {foreachVariableString} = {collectionVariableName}[{indexString}];", options: parseOption), SyntaxFactory.ElasticMarker);
```

Comment:
please consider wrapping lines that go past github (130 characters)

Reply:
sure

Reply:
also, please do not parse to make strings.  just construct the tree manually using SyntaxFactory or SyntaxGenerator.

Reply:
why? I dont see any benefit going tree generation here. and much straight-forward on what I am getting here.

Reply:
Well, for one thing, you then had to fix things up by adding elastic annotations.  

For another, we already have really good APIs for this.  Like SyntaxGenerator.LocalDeclarationStatement.

For another, because your approach hardcodes in 'var' instead of respecting the user's setting.

And, finally, because no other code does this.  please do not add code to the IDE that needlessly behaves differently from every other refactoring/fix we've created so far.